packages/multichain/src/scope/types.ts

```diff
@@ -0,0 +1,77 @@
+import {
+  isCaipNamespace,
+  isCaipChainId,
+  parseCaipChainId,
+} from '@metamask/utils';
+import type {
+  CaipChainId,
+  CaipReference,
+  CaipAccountId,
+  KnownCaipNamespace,
+  CaipNamespace,
+  Json,
+} from '@metamask/utils';
+
+// These External prefixed types represent the CAIP-217
+// Scope and ScopeObject as defined in the spec.
+export type ExternalScopeString = CaipChainId | CaipNamespace;
+export type ExternalScopeObject = Omit<ScopeObject, 'accounts'> & {
+  references?: CaipReference[];
+  accounts?: CaipAccountId[];
+};
+export type ExternalScopesObject = Record<
+  ExternalScopeString,
+  ExternalScopeObject
+>;
+
+// These non-prefixed types represent CAIP-217 Scope and
+// ScopeObject as defined by the spec but without
+// namespace-only Scopes (except for "wallet") and without
+// the `references` array of CAIP References on the ScopeObject.
+// These deviations from the spec are necessary as MetaMask
+// does not support wildcarded Scopes, i.e. Scopes that only
+// specify a namespace but no specific reference.
+export type ScopeString = CaipChainId | KnownCaipNamespace.Wallet;
+export type ScopeObject = {
+  methods: string[];
+  notifications: string[];
+  accounts: CaipAccountId[];
+  rpcDocuments?: string[];
+  rpcEndpoints?: string[];
+};
+export type ScopesObject = Record<CaipChainId, ScopeObject> & {
+  [KnownCaipNamespace.Wallet]?: ScopeObject;
+};
```

Comment:
Similar as above, would it be worth encapsulating these notes in JSDoc? Here is an attempt at expanding upon these slightly. If this is not a correct understanding then feel free to adjust.

``` suggestion
/**
 * Represents a `scopeString` as defined in
 * [CAIP-217](https://chainagnostic.org/CAIPs/caip-217), with the exception that
 * CAIP namespaces (aside from "wallet") are disallowed. This is necessary as
 * MetaMask does not support "wildcarded" scopes in CAIP-25 requests, i.e.
 * scopes that only specify a CAIP namespace but no specific reference (e.g.
 * "eip155" vs. "eip155:1").
 */
export type ScopeString = CaipChainId | KnownCaipNamespace.Wallet;

/**
 * Represents a `scopeObject` as defined in
 * [CAIP-217](https://chainagnostic.org/CAIPs/caip-217), with the exception that
 * the `references` property is disallowed. This is necessary as MetaMask does
 * not support "wildcarded" scopes in CAIP-25 requests, i.e. scopes that only
 * specify a CAIP namespace but no specific reference (e.g. "eip155" vs.
 * "eip155:1").
 */
export type ScopeObject = {
  methods: string[];
  notifications: string[];
  accounts: CaipAccountId[];
  rpcDocuments?: string[];
  rpcEndpoints?: string[];
};

/**
 * Represents a keyed `scopeObject` as defined in
 * [CAIP-217](https://chainagnostic.org/CAIPs/caip-217), with the exception that
 * `scopeObject`s do not contain `references`. This is necessary as MetaMask
 * does not support "wildcarded" scopes in CAIP-25 requests, i.e. scopes that
 * only specify a CAIP namespace but no specific reference (e.g. "eip155" vs.
 * "eip155:1").
 */
export type ScopesObject = Record<CaipChainId, ScopeObject> & {
  [KnownCaipNamespace.Wallet]?: ScopeObject;
};
```

Reply:
Actually it seems like in addition to the deviations listed above, `accounts` is also a required property in our version whereas in the spec it is optional. Is there a reason for this?

Reply:
> // These non-prefixed types represent CAIP-217 Scope and
> // ScopeObject as defined by the spec but without
> // namespace-only Scopes (except for "wallet") and without
> // the `references` array of CAIP References on the ScopeObject.
> // These deviations from the spec are necessary as MetaMask
> // does not support wildcarded Scopes, i.e. Scopes that only
> // specify a namespace but no specific reference.

Ok so I think Jiexi's original comment is confusing since it suggests that we don't allow CAIP-25 requests that use the following format:
```
...
"method": "wallet_createSession",
  "params": {
    "requiredScopes": {
      "eip155": {
        "references": ["1", "137"],
        "methods": ["eth_sendTransaction", "eth_signTransaction", "eth_sign", "get_balance", "personal_sign"],
        "notifications": ["accountsChanged", "chainChanged"]
      },
...
```
Where the scopeString is only a namespace and then the `references` are specified in an array. 

In fact we do allow this request format. What I have to assume Jiexi's original comment was meaning to indicate is that in our **internal representations** of this request/session we flatten out this format such that it becomes:

```
...
"method": "wallet_createSession",
  "params": {
    "requiredScopes": {
      "eip155:1": {
        "methods": ["eth_sendTransaction", "eth_signTransaction", "eth_sign", "get_balance", "personal_sign"],
        "notifications": ["accountsChanged", "chainChanged"]
      },
       "eip155:137": {
        "methods": ["eth_sendTransaction", "eth_signTransaction", "eth_sign", "get_balance", "personal_sign"],
        "notifications": ["accountsChanged", "chainChanged"]
      },
}
...
```

This is the reason for the two sets of types, those prefixed with "External" and those without:
- `ExternalScopeString` & `ScopeString`
- `ExternalScopeObject` & `ScopeObject`
- `ExternalScopesObject` & `ScopesObject`

Given this I'm going to rephrase the JSDoc comments slightly and push them.

Reply:
☝️ is also the explanation for this question👇 

> Actually it seems like in addition to the deviations listed above, accounts is also a required property in our version whereas in the spec it is optional. Is there a reason for this?

Notice that the `ExternalScopeObject`:
```
export type ExternalScopeObject = Omit<ScopeObject, 'accounts'> & {
  references?: CaipReference[];
  accounts?: CaipAccountId[];
};
```
resets the `accounts` property as optional. This is because the `ExternalScopeObject` is the type we expect in the CAIP-25 request itself for which, as you mentioned, the `accounts` property is optional. 

in the **internal** `ScopeObject` however this property is not optional.

Perhaps using an `internal` prefix would be less confusing?

Reply:
>Perhaps using an internal prefix would be less confusing?

Yeah, I think so. Either that or something like `Canonical` or `Normalized`. Maybe `Internal` is best though.

Reply:
Yeah, I think using `Internal`/`External` (or `Unofficial`/`Official`, or `MetaMask`/`Canonical`, or something) would help. Otherwise it gives the impression that the spec is deviating from something we are setting, when it's the other way around.

And updating the JSDoc to represent what we are actually doing would be good too.

Reply:
Added JSDocs here: https://github.com/MetaMask/core/pull/4784/commits/9c40d696318f7c031278ca3233f1ca50b4d9a3e2

And added `Internal` prefix to these types here: https://github.com/MetaMask/core/pull/4784/commits/71f254a16d58ea81640e41f6d32acc9b83481f86